### PR TITLE
Ensure encrypted events are not handled twice

### LINKED ISSUE #266: EncryptedEventBroker may emit twice for a single event when two users are syncing
Like a race somewhere.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #266 

This should address some problems with the encrypted event tracking

### CODE REVIEW

File: src/components/encryption.ts [hunk 1]
```diff
@@ -86,16 +86,18 @@ export class EncryptedEventBroker {
         private onEphemeralEvent: ((event: EphemeralEvent) => void)|undefined,
         private getIntent: (userId: string) => Intent
         ) {
+        if (this.onEphemeralEvent) {
```
Comment (Half-Shot) [inline src/components/encryption.ts:89]: If we're not handling presence, we don't need a cleanup timer.

File: src/components/encryption.ts [hunk 2]
```diff
@@ -196,10 +196,23 @@ export class EncryptedEventBroker {
...
-        log.debug(`Handling ${event.getId()} through sync`);
-        this.onEvent(event.event);
+        this.handledEvents.add(`${event.room_id}:${event.event_id}`);
+
+        log.debug(`Handling ${event.event_id} through sync`);
+        this.onEvent(event);
+
+        // Delete the event from the pending list
+        this.eventsPendingSync.delete(event.event_id);
+        let index = this.eventsPendingAS.findIndex((e) => e.event_id === event.event_id);
```
Comment (Half-Shot) [inline src/components/encryption.ts:211]: It's possible we might get multiple sync events stored.

File: src/components/encryption.ts [hunk 3]
```diff
@@ -267,13 +280,23 @@ export class EncryptedEventBroker {
...
         if (this.onEphemeralEvent) {
             matrixClient.on("RoomMember.typing", (event: TypingEvent) => this.onTyping(userId, event));
             matrixClient.on("Room.receipt", (event: ReadReceiptEvent) => this.onReceipt(userId, event));
             matrixClient.on("User.presence", (event: PresenceEvent) => this.onPresence(event));
         }
-        const filter = new Filter(userId);
-        filter.setDefinition(SYNC_FILTER);
+        else {
+            filter.definition.presence = {
+                types: ["not.a.real.type"],
```
Comment (Half-Shot) [inline src/components/encryption.ts:292]: The spec doesn't have a surefire way to turn off presence, but we can trick it by doing this.

File: src/components/encryption.ts (hunk 1 above)
Comment (benparsons) [reply src/components/encryption.ts:89]: this could be a comment in the code, not here

File: src/components/encryption.ts [hunk 4]
```diff
@@ -86,16 +86,18 @@ export class EncryptedEventBroker {
         private onEphemeralEvent: ((event: EphemeralEvent) => void)|undefined,
         private getIntent: (userId: string) => Intent
         ) {
+        if (this.onEphemeralEvent) {
+            this.presenceCleanupInterval = setInterval(() => {
+                const ts = Date.now() - 30000;
```
Comment (benparsons) [inline src/components/encryption.ts:91]: 30s wants to be a config var?

File: src/components/encryption.ts (hunk 3 above)
Comment (benparsons) [reply src/components/encryption.ts:292]: This can be an inline comment

File: src/components/encryption.ts [hunk 5]
```diff
@@ -196,10 +196,23 @@ export class EncryptedEventBroker {
...
+        this.handledEvents.add(`${event.room_id}:${event.event_id}`);
+
+        log.debug(`Handling ${event.event_id} through sync`);
+        this.onEvent(event);
+
+        // Delete the event from the pending list
+        this.eventsPendingSync.delete(event.event_id);
+        let index = this.eventsPendingAS.findIndex((e) => e.event_id === event.event_id);
+        do {
+            this.eventsPendingAS.splice(index, 1);
```
Comment (benparsons) [inline src/components/encryption.ts:213]: Would all of this be possible with `this.eventsPendingAS = this.eventsPendingAS.filter((e) => e.event_id !== event.event_id)`?

**Review by benparsons — CHANGES_REQUESTED** [review]
questions

File: src/components/encryption.ts (hunk 5 above)
Comment (Half-Shot) [reply src/components/encryption.ts]: Yeah, I prefer that.

File: src/components/encryption.ts (hunk 4 above)
Comment (Half-Shot) [reply src/components/encryption.ts]: We should make it a const but I don't think there is any value in it, it's just a cache lifetime.

**Review by benparsons — APPROVED** [review]
lgtm

**Pull request merged**